Pendle | Ethereum | 2025.02.19 (#99)
deficollective/defiscan: src/content/protocols/pendle.md

The SY tokens themselves can be paused by the _Pendle Team_. Calling `pause` locks the yield bearing assets in Pendle instantly without timelock protection. This could lead to permanent or temporary _loss of user funds_. Additionally, during a paused SY token, trading between the associated YT and PT is paused.

The `PendleYieldContractFactory`, controlled by the `Governance` multisig account, can change the interest fee on a market instantaneously up to 20% leaving them no opportunity to exit their positions. The fee rate directly impacts both existing positions and newly created ones. This could lead to _loss of unclaimed yield.

The `PendleCommonSYFactory` is used to create new markets and instantiate SY contracts. The implementation code used by the Factory to deploy new SY tokens can be updated by `Pendle: Deployer 1`, an EOA. Importantly, existing SY token contracts and user funds already deposited in the protocol are not affected by this vulnerability, as the risk applies only to new SY token deployments.

> ### Pendle Swap Module

Reviewer: the pendle swap module section can be further improved

External tokens → PT+YT (`mintPyFromToken`)  
Users can also convert external tokens (such as ETH or USDe) in a single transaction. The `PendleRouterV4` first converts these tokens to SY via a deposit operation, then transfers the SY to the `PendleYieldToken` contract which proceeds to issue PT and YT tokens in the same manner.

In both cases, the `PendleYieldToken` contract serves as the vault for SY tokens and coordinates the creation of equal quantities of PT and YT. The `PendlePrincipalToken` contract does not hold any assets directly, while the `PendleYieldToken` contract retains all SY tokens as collateral.

Governance can interrupt this process by pausing SY operations, thus blocking deposits, withdrawals, and transfers and preventing the creation of new PT/YT.

![Direct Minting](./diagrams/pendle-v2-direct-minting.png)

> 

Reviewer: @CookingCryptos can you add a unwrapping section SY -> Underlying

Author: I added that

SY → Underlying Assets (`redeemSyToToken`)  
Users can convert SY tokens back to the underlying assets through the `PendleRouterV4`. The router calls the respective SY contract's redeem function, which unwraps the SY token and returns the underlying asset (such as ETH, USDC, or wstETH) directly to the user. This process allows users to exit the Pendle ecosystem completely and reclaim their original assets.

Buying YT:
Users can acquire YT tokens without holding PT via `swapExactTokenForYt`. The router converts the external tokens to SY tokens, then temporarily borrows additional SY from the `PendleMarketV3` contract. All SY are used to mint both PT and YT (as they must be created in equal quantities), but only the YT are kept for the user. The PT are immediately sold back to the market to repay the borrowed SY. A portion of SY is sent as fees to the `Treasury`.

Selling YT:
When selling YT via `swapExactYtForToken`, the router borrows an equivalent amount of PT from the market, combines them with the user's YT to redeem SY (as both tokens are needed to unlock the underlying asset), then returns a portion of the SY to the market to repay the borrowed PT. The remaining SY are converted to the desired output token. As with buying, fees are collected and sent to the `Treasury`.

These flash swap mechanisms allow users to gain exposure to yield alone (YT) without needing to manage the principal component (PT).

`Governance` can `pause` SY token transfers, disabling the buying and selling of YT through this mechanism.

## PENDLE incentives and Fees

> 

Reviewer: @CookingCryptos can you write here more, maybe expand a little compared to the section above in the upgradeability which is less technical

x: "https://x.com/pendle_fi"
github: ["https://github.com/pendle-finance"]
defillama_slug: ["pendle"]
chain: "Ethereum"
stage: "O"
reasons: ["Unverified Contracts"]
risks: ["L", "H", "M", "H", "H"]
author: ["CookingCryptos, PaulDeFi"]
submission_date: "2025-02-19"
publish_date: "1970-01-01"
update_date: "1970-01-01"
---

# Summary

Pendle is a decentralized, permissionless yield-trading protocol that enables users to manage and optimize their yield exposure. By tokenizing yield-bearing assets into standardized yield tokens (SY), Pendle allows these assets to be split into principal tokens (PT) and yield tokens (YT). This separation facilitates various strategies, including earning fixed yields, speculating on future yield changes, and providing liquidity to earn additional rewards. Pendle's Automated Market Maker (AMM) is specifically designed to handle the trading of these time-decaying assets, ensuring efficient and flexible yield management within the decentralized finance ecosystem.

# Ratings

## Chain

Pendle is deployed on various chains. This review is based on the Ethereum mainnet deployment of the protocol.

> Chain score: Low

## Upgradeability

> The Pendle V2 protocol can be analyzed in a number of logical modules: _Yield Tokenization_, _Pendle AMM Swap_, _vePENDLE, fees and incentives_ and _Governance_. Each module has upgradeable parts. Overall, these vectors could result in the _loss of user funds_, _loss of unclaimed yield_ or otherwise _materially affect the expected performance_ of the protocol. All the control vectors are hold by accounts with insufficient decentralisation and no onchain Governance exists to this date. See the [Security Council Table](#security-council). With the current setup the _Upgradeability Score_ is _High_. 

Reviewer: ```suggestion
The Pendle V2 protocol can be analyzed in a number of logical modules: _Yield Tokenization_, _Pendle AMM Swap_, _vePENDLE, fees and incentives_ and _Governance_. Each module has upgradeable parts. Overall, these vectors could result in the _loss of user funds_, _loss of unclaimed yield_ or otherwise _materially affect the expected performance_ of the protocol. All the control vectors are held by accounts with insufficient decentralisation and no onchain Governance exists to this date. See the [Security Council Table](#security-council). With the current setup the _Upgradeability Score_ is _High_. 
```

> The permission to upgrade the protocol follows a hierarchical structure. The `Governance`, a 2/4 multisig that serves as the `DEFAULT_ADMIN_ROLE` of the `PendleGovernanceProxy`, has ultimate authority over the protocol's contracts. It directly controls the `ProxyAdmin` (which manages all `TransparentUpgradeableProxy` contracts) and also has direct control over various other contracts including voting mechanisms, market factories, and cross-chain messaging. The `DevMultisig`, a 2/3 multisig, controls fee distribution systems and reward distribution mechanisms like `PendleFeeDistributorV2` and `PendleMultiTokenMerkleDistributor`.

Author: The Pendle V2 protocol can be analyzed in a number of logical modules: _Yield Tokenization_, _Pendle AMM Swap_, _vePENDLE, fees and incentives_ and _Governance_. Each module has upgradeable parts. Overall, these vectors could result in the _loss of user funds_, _loss of unclaimed yield_ or otherwise _materially affect the expected performance_ of the protocol. All the control vectors are hold by accounts with insufficient decentralisation and no onchain Governance exists to this date. See the [Security Council Table](#security-council). With the current setup the _Upgradeability Score_ is _High_. 

The permission to upgrade the protocol follows a hierarchical structure. The `Governance`, a 2/4 multisig that serves as the `DEFAULT_ADMIN_ROLE` of the `PendleGovernanceProxy`, has ultimate authority over the protocol's contracts. It directly controls the `ProxyAdmin` (which manages all `TransparentUpgradeableProxy` contracts) and also has direct control over various other contracts including voting mechanisms, market factories, and cross-chain messaging. The `DevMultisig`, a 2/3 multisig, controls fee distribution systems and reward distribution mechanisms like `PendleFeeDistributorV2` and `PendleMultiTokenMerkleDistributor`.

Does this text replace that one, is it correct?

> The permission to upgrade the protocol follows a hierarchical structure. The `Governance`, a 2/4 multisig that serves as the `DEFAULT_ADMIN_ROLE` of the `PendleGovernanceProxy`, has ultimate authority over the protocol's contracts. It directly controls the `ProxyAdmin` (which manages all `TransparentUpgradeableProxy` contracts) and also has direct control over various other contracts including voting mechanisms, market factories, and cross-chain messaging. The `DevMultisig`, a 2/3 multisig, controls fee distribution systems and reward distribution mechanisms like `PendleFeeDistributorV2` and `PendleMultiTokenMerkleDistributor`.

Reviewer: This goes into governance

> The permission to upgrade the protocol follows a hierarchical structure. The `Governance`, a 2/4 multisig that serves as the `DEFAULT_ADMIN_ROLE` of the `PendleGovernanceProxy`, has ultimate authority over the protocol's contracts. It directly controls the `ProxyAdmin` (which manages all `TransparentUpgradeableProxy` contracts) and also has direct control over various other contracts including voting mechanisms, market factories, and cross-chain messaging. The `DevMultisig`, a 2/3 multisig, controls fee distribution systems and reward distribution mechanisms like `PendleFeeDistributorV2` and `PendleMultiTokenMerkleDistributor`.

This we dont need

> The Pendle V2 protocol is fully upgradeable through multiple proxy patterns including `ERC1967/UUPS` proxies for most core contracts and `TransparentUpgradeableProxy` contracts managed by a central `ProxyAdmin`. This comprehensive upgradeability capability can potentially result in loss of user funds, theft of unclaimed rewards, and significant changes to the protocol's economic parameters and performance.

> The permission to upgrade the protocol follows a hierarchical structure. The `Governance`, a 2/4 multisig that serves as the `DEFAULT_ADMIN_ROLE` of the `PendleGovernanceProxy`, has ultimate authority over the protocol's contracts. It directly controls the `ProxyAdmin` (which manages all `TransparentUpgradeableProxy` contracts) and also has direct control over various other contracts including voting mechanisms, market factories, and cross-chain messaging. The `DevMultisig`, a 2/3 multisig, controls fee distribution systems and reward distribution mechanisms like `PendleFeeDistributorV2` and `PendleMultiTokenMerkleDistributor`.

Reviewer: This remains

> The Pendle V2 protocol can be analyzed in a number of logical modules: Yield Tokenization, Pendle AMM Swap, vePENDLE, fees and incentives and Governance. Each module has upgradeable parts. Overall, these vectors could result in the loss of user funds, loss of unclaimed yield or otherwise materially affect the expected performance of the protocol. All the control vectors are hold by accounts with insufficient decentralisation and no onchain Governance exists to this date. See the [Security Council Table](https://github.com/deficollective/defiscan/pull/99#security-council). With the current setup the Upgradeability Score is High.

> ### Core Protocol Components

Author: This is in Governance now

> The `BaseSplitCodeFactoryContract` creates large contracts by splitting their bytecode into two separate fragments that function together as a single logical entity. The resulting fragment contracts remain unverified on Etherscan, creating significant transparency issues as their source code is inaccessible to the public. This contract is controlled by the EOA "Pendle: Deployer 1" and is used to deploy contracts like `PendleMarketFactoryV3`.

Reviewer: move this to Reviewer Notes section

> The `ProxyAdmin` contract controls all `TransparentUpgradeableProxy` contracts in the protocol through functions including `upgrade`, `upgradeAndCall`, and `changeProxyAdmin`. These functions enable complete replacement of contract logic while preserving state data, with no timelock protection.

Reviewer: This is too much technical detail, and readers wouldn't know which contracts (AMM? YT? PT?) are of type `TransparentUpgradeableProxy`

Reviewer: and saying it can be upgraded is enough, listing exact functions like  `upgrade`, `upgradeAndCall`, and `changeProxyAdmin` is too much detail

The `upgradeToAndCall` function, available across the system's proxy contracts, permits upgrading a contract and executing initialization code atomically in a single transaction. Among all protocol contracts, only the `PENDLE` token implements a timelock mechanism (7 days via `initiateConfigChanges`). All other contracts can be upgraded instantaneously without delay or community review.

### Yield Tokenization Infrastructure

> The Yield Tokenization infrastructure presents significant upgradeability risks through multiple interconnected contracts. The `PendleCommonSYFactory`  is controlled by "Pendle: Deployer 1", an EOA. This contract can register arbitrary implementation code for new SY tokens via `setSYCreationCode`, creating a critical vulnerability where malicious code could be deployed for all future yield-bearing assets.

Reviewer: mention explicitly that existing contracts and user funds are not affected

The `PendleYieldContractFactory`, controlled by the `Governance` , can modify economic parameters without timelock via `setInterestFeeRate` and `setTreasury`. These functions allow instant changes to protocol fees (up to 20%) and redirection of revenue streams without warning users, leaving them no opportunity to exit their positions.

> The SY tokens themselves `PendleERC4626SYV2` (example SY contract) contain `pause` and `unpause` functions that can be triggered by any address with appropriate permissions in the `PendleGovernanceProxy`. This creates an additional upgradeability risk where user assets could be instantly frozen through governance actions without timelock protection.

Reviewer: ```suggestion
The SY tokens themselves (`PendleERC4626SYV2`) contain `pause` and `unpause` functions that can be triggered by the _Pendle Team_. Calling `pause` locks deposited user assets instantly without timelock protection. This could lead to permanent _loss of user funds_.
```

> ### Pendle Swap System

Reviewer: ```suggestion
### Pendle Swap Module
```

Reviewer: we speak about modules

> The PendleSwap component employs a custom selector-based router proxy `PendleRouterV4` that directs calls to seven specialized implementations. Unlike standard upgradeable proxies, PendleRouterV4's routing infrastructure is controlled by `ActionStorageV4`, whose owner is permanently set to the zero address, making its configuration immutable and eliminating upgradeability risks for this specific component.

Reviewer: > The PendleSwap component employs a custom selector-based router proxy `PendleRouterV4`

The `PendleRouterV4` routes all user actions not just swap, but also tokenization

---
protocol: "Pendle"
website: "https://www.pendle.finance/"
x: "https://x.com/pendle_fi"
github: ["https://github.com/pendle-finance"]
defillama_slug: ["pendle"]
chain: "Ethereum"
stage: "O"
reasons: ["Unverified Contracts"]
risks: ["L", "H", "L", "H", "H"]
author: ["CookingCryptos, PaulDeFi"]
submission_date: "2025-02-19"
publish_date: "1970-01-01"
update_date: "1970-01-01"
---

# Summary

Pendle is a yield-trading protocol that enables users to manage and optimize their yield exposure. By tokenizing yield-bearing assets into standardized yield tokens (SY), Pendle allows these assets to be split into principal tokens (PT) and yield tokens (YT). This separation facilitates various strategies, including earning fixed yields, speculating on future yield changes, and providing liquidity to earn additional rewards. Pendle's Automated Market Maker (AMM) is specifically designed to handle the trading of these time-decaying assets.

# Ratings

## Chain

Pendle is deployed on various chains. This review is based on the Ethereum mainnet deployment of the protocol.

> Chain score: Low

## Upgradeability

> The Pendle V2 protocol can be analyzed in a number of logical modules: _Yield Tokenization_, _Pendle AMM Swap_, _vePENDLE, fees and incentives_ and _Governance_. Each module has upgradeable parts. Overall, these vectors could result in the _loss of user funds_, _loss of unclaimed yield_ or otherwise _materially affect the expected performance_ of the protocol. All the control vectors are hold by accounts with insufficient decentralisation and no onchain Governance exists to this date. See the [Security Council Table](#security-council). With the current setup the _Upgradeability Score_ is _High_. 

Reviewer: ```suggestion
The Pendle V2 protocol can be analyzed in a number of logical modules: _Yield Tokenization_, _Pendle AMM Swap_, _vePENDLE, fees and incentives_ and _Governance_. Each module has upgradeable parts. Overall, these vectors could result in the _loss of user funds_, _loss of unclaimed yield_ or otherwise _materially affect the expected performance_ of the protocol. All the control vectors are held by accounts with insufficient decentralisation and no onchain Governance exists to this date. See the [Security Council Table](#security-council). With the current setup the _Upgradeability Score_ is _High_. 
```

### Core Protocol Components

> Yield Tokenization is the foundation of Pendle, converting yield-bearing assets into standardized yield (SY) tokens which are then split into Principal Tokens (PT) and Yield Tokens (YT). This separation allows users to trade yield exposure independently.

Reviewer: move this 

> Yield Tokenization is the foundation of Pendle, converting yield-bearing assets into standardized yield (SY) tokens which are then split into Principal Tokens (PT) and Yield Tokens (YT). This separation allows users to trade yield exposure independently.

at the beginning of the _### Yield Tokenization_ Infrastructure section

Delete the _### Core Protocol Components_ section and the two remaining sentences.

> Upgradeability score: High

## Autonomy

Pendle deployment on Ethereum does not depend on any dependency. Please read the [Reviewer Notes](#reviewer-notes) to informed about the dependency risk on other chain deployments.

> Autonomy score: Low

## Exit Window

With the exception of the `PENDLE` token which implements a 7-day delay through the `initiateConfigChanges` function, most other protocol contracts can be upgraded or modified instantly without any delay or warning to users. 

> Exit window score: High

## Accessibility

Users can only access Pendle through a single user interface: [app.pendle.finance](https://app.pendle.finance). Currently, there is no backup solution in case the interface is shut down or users are censored.

> Accessibility score: High

## Conclusion

The Pendle V2 protocol falls into the N/A category and not Stage 0 due to the `sySwapper` contract which is not verified on public block explorers.

Despite this classification, our analysis shows that Pendle V2 protocol receives High centralization risk scores for Upgradeability, Exit Window, and Accessibility dimensions, and a low score for Autonomy.

> To reach Stage 0, Pendle should first verify all its contracts on Etherscan. Then, to progress toward Stage 1, the protocol would need to: 1) assign permissions to an account that suffices the _Security Council requirements_ with at least 7 signers, a 51% threshold, and at least 50% publicly identified external signers; 2) implement timelock mechanisms for critical functions; and 3) improve accessibility by publishing the frontend code to IPFS or developing alternative interfaces.

Author: To reach Stage 0, Pendle should first verify all its contracts on Etherscan. Then, to progress toward Stage 1, the protocol would need to: 1) assign permissions to an account that suffices the _Security Council requirements_ with at least 7 signers, a 51% threshold, and at least 50% publicly identified external signers Or 2) implement timelock mechanisms for critical functions.
And 3) improve accessibility by publishing the frontend code to IPFS or developing alternative interfaces.

`PendleYieldContractFactory` enables the creation of Principal Token (PT) and Yield Token (YT) contracts through its `createYieldContract` function, which is also permissionless. This function can be called directly by users or programmatically via the `PendleCommonPoolDeployHelperV2` as part of a full pool deployment process which also deploys the market.

![Market Creation](./diagrams/pendle-v2-market-creation.png)

## Yield Tokenization Process

### Direct Minting

SY → PT+YT (`mintPyFromSy`)  
Users with existing SY tokens can convert them directly into PT and YT through the `PendleRouterV4`. The router passes the request to the `PendleYieldToken` contract, which receives the SY tokens and then calls its `mintPY` function on the YT contract. The `PendleYieldToken` contract mints YT tokens and requests the `PendlePrincipalToken` contract to mint the corresponding amount of PT tokens via `mintByYT`.

External tokens → PT+YT (`mintPyFromToken`)  

> Users can also convert external tokens (such as ETH or USDe) in a single transaction. The `PendleRouterV4` first converts these tokens to SY via a deposit operation, then transfers the SY to the `PendleYieldToken` contract which proceeds to issue PT and YT tokens in the same manner.

Author: It was for showing that there are multiple ways to mint YT and PT. I changed for that :

"PT and YT tokens are minted via the `PendleRouterV4`, fundamentally by converting Standardized Yield (SY) tokens. Users with SY can directly convert them to PT and YT using `mintPyFromSy`, where the `PendleYieldToken` contract receives SY, mints YT, and triggers `PendlePrincipalToken` to mint PT. Alternatively, `mintPyFromToken` allows users to start with external assets (e.g., ETH), which the router first converts to SY before following the same SY-to-PT/YT minting pathway."

SY → Underlying Assets (`redeemSyToToken`)  
Users can convert SY tokens back to the underlying assets through the `PendleRouterV4`. The router calls the respective SY contract's redeem function, which unwraps the SY token and returns the underlying asset (such as ETH, USDC, or wstETH) directly to the user. This process allows users to exit the Pendle ecosystem completely and reclaim their original assets.

## Providing Liquidity (LP)  

Users can add liquidity through the `addLiquiditySingleToken` function of `PendleRouterV4`, which delegates the operation to the `ActionAddRemoveLiqV3` contract. This process converts the deposited tokens to SY, uses a portion to acquire PT via `swapSyForExactPt`, and issues LP tokens representing a position in the PT/SY pool. Fees are collected and transferred to the Treasury.

Liquidity removal is performed via `removeLiquiditySingleToken`. The LP tokens are burned, the user receives their proportional share of SY and PT, the PT are converted to SY via `swapExactPtForSy`, and the SY are converted to underlying tokens. During both addition and removal, the `PendleGaugeController` contract distributes PENDLE rewards to the market and the user.

> ![LP to Market](./diagrams/pendle-v2-lp.png)

Author: I don't see any multisig here.

Users swaps are the next module

## Router System and Action Modules

The `PendleRouterV4` functions as the protocol's central interface hub, directing user interactions to specialized execution modules. This architecture employs a selector-based routing mechanism where each function call is mapped to one of seven dedicated action modules, each responsible for specific protocol operations.

The primary modules include `ActionAddRemoveLiqV3` for liquidity management in Pendle pools, `ActionSwapPTV3` for Principal Token exchange operations, and `ActionSwapYTV3` for Yield Token transactions. Supporting modules such as `ActionMiscV3` handle yield redemption, token conversions, and position exits, while `ActionSimple` provides core utility functions. Advanced functionalities are enabled through `ActionAggregator` for batched operations and `ActionLimitOrder` for conditional trades.

> 

Author: I think this paragraph on the action modules is quite important for understanding the router's special architecture and how functionalities are organized. But if you feel it's not worth keeping, I'll remove it straight away.

## Limit Order System

The protocol implemented the `PendleLimitRouter` contract which enables limit orders of PT and YT. 

Users can create orders that execute only when specific price conditions are met and can cancel their orders individually or in batches. The system uses EIP-712 signatures for order validation and employs logarithmic calculations for fee determination. These fee parameters are configured via `setLnFeeRateRoots` by either `Governance` or the `hardwareDeployer`, while collected fees are directed to `Treasury1`.

## PENDLE incentives and Fees

### Fee Collection and Distribution Path

The protocol implements a multi-stage fee distribution architecture with distinct flows for different market types. Fee revenue generated through market operations flows into two separate Treasury multisig accounts: `Treasury1` (2/6 multisig) processes fees from pools created before October 8, 2024, while `Treasury2` (2/5 multisig) handles those from newer pools.

> From these Treasury accounts, tokens follow a standardized conversion path. The `hardwareDeployer` EOA serves as an intermediary conversion agent, receiving tokens from both Treasury multisigs. This EOA transforms various token types into ETH through the `PendleRouterV4`, then directs this ETH to the `PendleFeeDistributorV2Proxy` for final distribution.

Reviewer 2: This is an EOA, I would emphasize that nothing guarrentees that it is going to distribute the fees

## PENDLE incentives and Fees

### Fee Collection and Distribution Path

The protocol implements a multi-stage fee distribution architecture with distinct flows for different market types. Fee revenue generated through market operations flows into two separate Treasury multisig accounts: `Treasury1` (2/6 multisig) processes fees from pools created before October 8, 2024, while `Treasury2` (2/5 multisig) handles those from newer pools.

From these Treasury accounts, tokens follow a standardized conversion path. The `hardwareDeployer` EOA serves as an intermediary conversion agent, receiving tokens from both Treasury multisigs. This EOA transforms various token types into ETH through the `PendleRouterV4`, then directs this ETH to the `PendleFeeDistributorV2Proxy` for final distribution.

The `PendleFeeDistributorV2Proxy` functions as the distribution contract for the collected fees that are rewarded to vePENDLE holders for their respective voting power. Under the `DevMultisig` (2/3) administration, this contract employs merkle proofs through `setMerkleRootAndFund` to set the claims while simultaneously depositing funds for claiming.

> The `hardwareDeployer` EOA maintains additional protocol functionalities with its `GUARDIAN` and `ALICE` roles in the `PendleGovernanceProxy`. These capacities include fee rate modifications through `setLnFeeRate` and operational controls.

Reviewer: @CookingCryptos please elaborate more here or remove line 204, in the Ratings section we wrote about the impact of lowering fees that this reduces the yield for vePENDLE voters, this is technically too vague and at the same time it does not give much additional knowledge

In parallel, the legacy `PendleFeeDistributor` remains active under the control of an unidentified address (0xD9c9935f4BFaC33F38fd3A35265a237836b30Bd1). This contract continues to distribute USDC to users through a separate reward system not accountable to the current Treasury structure.

### PENDLE Emissions to LPs

> The PENDLE token emission system operates through an interconnected contract architecture with dynamic reward distribution.

Reviewer: @CookingCryptos  this sentence (line 210) does inform the reader not really. In our reports everything is about "interconnected contract architectures" - so this is just stating the obvious. Probably just remove this sentence, or have a introduction to PENDLE emissions to LPs (prime the reader for what is coming)

Author: The Pendle protocol distributes PENDLE tokens as incentives to liquidity providers (LPs) through a weighted voting system. vePENDLE holders can allocate their voting power across different pools.

The emission cycle begins when the `PendleLiquidityIncentivesMultisig` claims newly minted PENDLE through `claimLiquidityEmissions` based on a predetermined schedule. The PENDLE token's parameters are protected by a 7-day timelock via `initiateConfigChanges`. These tokens then flow to `Governance`, which funds the `PendleGaugeControllerMainchainUpg` through the `fundPendle` function.

vePENDLE holders lock their PENDLE in the `VotingEscrowPendleMainchain` contract and direct incentives by voting on pool allocations through `vote(pools, weights)` in the `PendleVotingControllerUpg`. This contract verifies vePENDLE balances and manages configuration parameters including emission rates (`setPendlePerSec`) and pool eligibility (`addPool`, `removePool`).

The `PendleVotingControllerUpg` sends voting results to the `PendleGaugeControllerMainchainUpg` via `updateVotingResults`, which then allocates PENDLE rewards to `PendleMarketV3` contracts dynamically.

> Finally, LPs can claim their accumulated rewards directly from `PendleMarketV3` contracts via `redeemRewards`, completing a system where incentives are continuously allocated in response to liquidity activity and governance decisions.

Reviewer: line 218, also mention that LPs receive their reward when closing their LP position

Author: Finally, LPs can claim their accumulated rewards directly from `PendleMarketV3` contracts via `redeemRewards` or automatically receive them when closing their LP position through `removeLiquiditySingleToken`.

### Points Rewards

The `PendleMultiTokenMerkleDistributor` manages token points system rewards for vePENDLE holders through a monthly cycle. The system takes a snapshot of vePENDLE balances on the 20th of each month, with token distributions occurring approximately one week later. The `DevMultisig` (2/3) controls this contract.

> The process uses the `setMerkleRoot` function to establish which addresses can claim which rewards, `upgradeToAndCall` to modify the implementation, and `transferOwnership` to transfer administrative control.

Reviewer: this sentence (line 224) does not make much sense, first part is good but

add additional sentences for the following information

`upgradeToAndCall` to modify the implementation, and `transferOwnership` to transfer administrative control.

> The `DevMultisig` can modify reward distributions through several mechanisms: changing which addresses receive rewards via `setMerkleRoot`, upgrading the contract implementation through `upgradeToAndCall`, and determining which token types are included in distributions. These parameters can be adjusted during the period between the monthly snapshot and distribution dates.

Reviewer: @CookingCryptos  duplicate information, consolidate with the thing above

Author: The process uses the `setMerkleRoot` function to establish which addresses can claim specific reward amounts, allowing the `DevMultisig` to control reward distribution between the monthly snapshot and payout dates. 
The contract implements the UUPS upgradeability pattern through `upgradeToAndCall`, enabling the `DevMultisig` to modify implementation logic while preserving claimed data, and can also determine which token types are included in distributions. Administrative control follows a two-step ownership transfer pattern via `transferOwnership`, requiring the pending owner to explicitly claim ownership for security.
